packages/next/src/server/lib/router-utils/route-types-shared.ts

```diff
@@ -0,0 +1,196 @@
+import path from 'path'
+import {
+  getRouteRegex,
+  type Group,
+} from '../../../shared/lib/router/utils/route-regex'
+
+export interface RouteInfo {
+  path: string
+  groups: { [groupName: string]: Group }
+}
+
+export interface RouteTypesManifest {
+  appRoutes: Record<string, RouteInfo>
+  pageRoutes: Record<string, RouteInfo>
+  layoutRoutes: Record<string, RouteInfo | (RouteInfo & { slots: string[] })>
+  /** Map of redirect source => RouteInfo */
+  redirectRoutes: Record<string, RouteInfo>
+  /** Map of rewrite source => RouteInfo */
+  rewriteRoutes: Record<string, RouteInfo>
+}
+
+// Convert a custom-route source string (`/blog/:slug`, `/docs/:path*`, ...)
+// into the bracket-syntax used by other Next.js route helpers so that we can
+// reuse `getRouteRegex()` to extract groups.
+export function convertCustomRouteSource(source: string): string {
+  // Handle catch-all (one or more / zero or more)  :param* / :param+
+  let out = source.replace(
+    /:([A-Za-z0-9_]+)\*/g,
+    (_m: string, name: string) => `[...${name}]`
+  )
+  out = out.replace(
+    /:([A-Za-z0-9_]+)\+/g,
+    (_m: string, name: string) => `[...${name}]`
+  )
+  // Optional catch-all  :param?
+  out = out.replace(
+    /:([A-Za-z0-9_]+)\?/g,
+    (_m: string, name: string) => `[[...${name}]]`
+  )
+  // Standard dynamic segment :param (ensure we don't convert already replaced ones)
+  out = out.replace(
+    /:([A-Za-z0-9_]+)/g,
+    (_m: string, name: string) => `[${name}]`
+  )
+  // Ensure leading slash
+  if (!out.startsWith('/')) out = '/' + out
+  return out
+}
+
+/**
+ * Extracts route parameters from a route pattern
+ */
+export function extractRouteParams(route: string) {
+  const regex = getRouteRegex(route)
+  return regex.groups
+}
+
+/**
+ * Determines if a route should be skipped during type generation
+ */
+export function shouldSkipRoute(filePath: string): boolean {
+  return (
+    filePath.includes('(..)') ||
```

Comment:
The characters are special but might be more accurate to assert on each segment, split with `path.sep` then assert if segment starts or ends with special characters here